File: backend/app/schemas/invoice.py
from datetime import date, datetime
from decimal import Decimal, InvalidOperation
import re

from pydantic import BaseModel, Field, field_validator

from app.models.invoice import InvoiceSource, InvoiceType
from app.schemas.validation_rules import STATE_CODE_BY_NAME, STATE_CODE_PATTERN, STATE_NAME_BY_LOWERCASE

INVOICE_NUMBER_PATTERN = re.compile(r'^\d{4}-\d{2}/\d{3}$')
DESCRIPTION_PATTERN = re.compile(r'^[A-Za-z0-9 ]+$')
HSN_SAC_PATTERN = re.compile(r'^\d{1,8}$')


def _decimal_places_within(value: float, max_places: int) -> bool:
    try:
        decimal_value = Decimal(str(value))
    except InvalidOperation:
        return False
    return abs(decimal_value.as_tuple().exponent) <= max_places


class InvoiceItemCreate(BaseModel):
    description: str = Field(min_length=1, max_length=20)
    hsn_sac: str = Field(min_length=1, max_length=8)
    quantity: float = Field(gt=0, le=5_000_000)
    rate: float = Field(ge=0)
    tax_rate: float = Field(ge=0, le=99.99)

    @field_validator('description')
    @classmethod
    def validate_description(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError('Description is required.')
        if not DESCRIPTION_PATTERN.fullmatch(cleaned):
            raise ValueError('Description should be alphanumeric and special characters are not allowed.')
        return cleaned

    @field_validator('hsn_sac')
    @classmethod
    def validate_hsn_sac(cls, value: str) -> str:
        cleaned = value.strip()
        if not HSN_SAC_PATTERN.fullmatch(cleaned):
            raise ValueError('HSN/SAC should contain only digits and be up to 8 digits.')
        return cleaned

    @field_validator('rate')
    @classmethod
    def validate_rate(cls, value: float) -> float:
        if not _decimal_places_within(value, 2):
            raise ValueError('Rate should have up to 2 decimal places.')
        return value

    @field_validator('tax_rate')
    @classmethod
    def validate_tax_rate(cls, value: float) -> float:
        if not _decimal_places_within(value, 2):
            raise ValueError('Tax Rate should have up to 2 decimal places.')
        return value


class InvoiceItemResponse(BaseModel):
    id: str
    description: str
    hsn_sac: str | None = None
    quantity: float
    rate: float
    tax_rate: float
    amount_before_tax: float
    cgst: float
    sgst_utgst: float
    total_tax_amount: float
    grand_total: float

    model_config = {'from_attributes': True}


class InvoiceCreate(BaseModel):
    client_id: str
    invoice_number: str = Field(min_length=11, max_length=11)
    invoice_date: date
    place_of_supply: str = Field(min_length=1, max_length=64)
    place_of_supply_code: str = Field(min_length=2, max_length=2)
    notes: str | None = None
    items: list[InvoiceItemCreate]

    @field_validator('client_id')
    @classmethod
    def validate_client_id(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError('Client is required.')
        return cleaned

    @field_validator('invoice_number')
    @classmethod
    def validate_invoice_number(cls, value: str) -> str:
        cleaned = value.strip()
        if not INVOICE_NUMBER_PATTERN.fullmatch(cleaned):
            raise ValueError('Invoice Number should be in format YYYY-YY/NNN.')

        start_year = int(cleaned[:4])
        next_year_short = int(cleaned[5:7])
        expected_next_year_short = (start_year + 1) % 100
        if next_year_short != expected_next_year_short:
            raise ValueError('Invoice Number year segment is invalid. Expected YYYY-(YYYY+1)/NNN format.')
        return cleaned

    @field_validator('place_of_supply')
    @classmethod
    def validate_place_of_supply(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError('Place of Supply is required.')
        normalized = STATE_NAME_BY_LOWERCASE.get(cleaned.lower())
        if normalized is None:
            raise ValueError('Please select a valid Indian state or union territory for Place of Supply.')
        return normalized

    @field_validator('place_of_supply_code')
    @classmethod
    def validate_place_of_supply_code(cls, value: str) -> str:
        cleaned = value.strip()
        if not STATE_CODE_PATTERN.fullmatch(cleaned):
            raise ValueError('Place of Supply Code should be exactly 2 digits.')
        return cleaned

    @field_validator('items')
    @classmethod
    def validate_items_not_empty(cls, value: list[InvoiceItemCreate]) -> list[InvoiceItemCreate]:
        if not value:
            raise ValueError('At least one item is required.')
        return value

    @field_validator('place_of_supply_code')
    @classmethod
    def validate_place_of_supply_code_match(cls, value: str, info) -> str:
        place_of_supply = info.data.get('place_of_supply')
        if not place_of_supply:
            return value
        expected_code = STATE_CODE_BY_NAME.get(place_of_supply.lower())
        if expected_code and value != expected_code:
            raise ValueError('Place of Supply Code does not match the selected Place of Supply.')
        return value


class InvoiceResponse(BaseModel):
    id: str
    client_id: str | None
    client_name: str | None = None
    invoice_number: str
    invoice_date: date
    place_of_supply: str | None = None
    place_of_supply_code: str | None = None
    gst_number: str | None
    type: InvoiceType
    subtotal: float
    gst_amount: float
    total_amount: float
    source: InvoiceSource
    notes: str | None
    original_file_path: str | None
    created_at: datetime
    items: list[InvoiceItemResponse]

    model_config = {'from_attributes': True}


class InvoiceListResponse(BaseModel):
    invoices: list[InvoiceResponse]
    count: int

File: frontend/lib/validation/invoice.ts
const INVOICE_NUMBER_PATTERN = /^\d{4}-\d{2}\/\d{3}$/;
const DESCRIPTION_PATTERN = /^[A-Za-z0-9 ]+$/;
const HSN_SAC_PATTERN = /^\d{4,15}$/;
const NUMBER_WITH_TWO_DECIMALS_PATTERN = /^\d+(\.\d{1,2})?$/;

export const MAX_QUANTITY = 5_000_000;

function hasMaxTwoDecimals(value: number): boolean {
  return Math.round(value * 100) === value * 100;
}

export function formatFinancialYearInvoicePrefix(dateValue: string): string {
  const parsedDate = new Date(dateValue);
  const year = Number.isNaN(parsedDate.getTime()) ? new Date() : parsedDate;
  const financialYearStart = year.getMonth() >= 3 ? year.getFullYear() : year.getFullYear() - 1;
  const financialYearEndShort = String((financialYearStart + 1) % 100).padStart(2, '0');
  return `${financialYearStart}-${financialYearEndShort}`;
}

export function buildDefaultInvoiceNumber(dateValue: string): string {
  return `${formatFinancialYearInvoicePrefix(dateValue)}/001`;
}

export function buildIncrementedInvoiceNumber(
  dateValue: string,
  latestInvoiceNumber?: string | null,
): string {
  const prefix = formatFinancialYearInvoicePrefix(dateValue);
  const candidate = (latestInvoiceNumber || '').trim();
  if (!INVOICE_NUMBER_PATTERN.test(candidate)) {
    return `${prefix}/001`;
  }

  const previousSerial = Number(candidate.slice(8));
  if (!Number.isFinite(previousSerial) || previousSerial <= 0 || previousSerial > 999) {
    return `${prefix}/001`;
  }

  const nextSerial = previousSerial >= 999 ? 1 : previousSerial + 1;
  return `${prefix}/${String(nextSerial).padStart(3, '0')}`;
}

export function sanitizeInvoiceNumberInput(value: string): string {
  return value.toUpperCase().replace(/[^0-9/-]/g, '').slice(0, 11);
}

export function validateInvoiceNumber(value: string): string | null {
  const trimmed = value.trim();
  if (!trimmed) return 'Invoice Number is required.';
  if (!INVOICE_NUMBER_PATTERN.test(trimmed)) {
    return 'Invoice Number should be in format YYYY-YY/NNN.';
  }

  const startYear = Number(trimmed.slice(0, 4));
  const nextYearShort = Number(trimmed.slice(5, 7));
  const expectedNextYearShort = (startYear + 1) % 100;
  if (nextYearShort !== expectedNextYearShort) {
    return 'Invoice Number year segment is invalid. Expected YYYY-(YYYY+1)/NNN format.';
  }
  return null;
}

export function sanitizeItemDescriptionInput(value: string): string {
  return value.replace(/[^A-Za-z0-9 ]/g, '').slice(0, 20);
}

export function validateItemDescription(value: string): string | null {
  const trimmed = value.trim();
  if (!trimmed) return 'Description is required.';
  if (!DESCRIPTION_PATTERN.test(trimmed)) {
    return 'Description should be alphanumeric and special characters are not allowed.';
  }
  return null;
}

export function sanitizeHsnSacInput(value: string): string {
  return value.replace(/\D/g, '').slice(0, 15);
}

export function validateHsnSac(value: string): string | null {
  const trimmed = value.trim();
  if (!trimmed) return 'HSN/SAC is required.';
  if (!HSN_SAC_PATTERN.test(trimmed)) return 'HSN/SAC should contain 4 to 15 digits.';
  return null;
}

export function sanitizeDecimalInput(value: string): string {
  const sanitized = value.replace(/[^0-9.]/g, '');
  const [whole = '', fraction = ''] = sanitized.split('.');
  if (!sanitized.includes('.')) return whole;
  return `${whole}.${fraction.slice(0, 2)}`;
}

export function validateQuantity(value: string): string | null {
  const trimmed = value.trim();
  if (!trimmed) return 'Quantity is required.';
  const numeric = Number(trimmed);
  if (!Number.isFinite(numeric) || numeric <= 0) return 'Quantity should be a valid number greater than 0.';
  if (numeric > MAX_QUANTITY) return 'Quantity should be up to 50,00,000.';
  return null;
}

export function validateRate(value: string): string | null {
  const trimmed = value.trim();
  if (!trimmed) return 'Rate is required.';
  if (!NUMBER_WITH_TWO_DECIMALS_PATTERN.test(trimmed)) {
    return 'Rate should be a number with up to 2 decimal places.';
  }
  const numeric = Number(trimmed);
  if (!Number.isFinite(numeric) || numeric < 0) return 'Rate should be a valid non-negative number.';
  if (!hasMaxTwoDecimals(numeric)) return 'Rate should have up to 2 decimal places.';
  return null;
}

export function validateTaxRate(value: string): string | null {
  const trimmed = value.trim();
  if (!trimmed) return 'Tax Rate is required.';
  if (!NUMBER_WITH_TWO_DECIMALS_PATTERN.test(trimmed)) {
    return 'Tax Rate should be a number with up to 2 decimal places.';
  }
  const numeric = Number(trimmed);
  if (!Number.isFinite(numeric) || numeric < 0) return 'Tax Rate should be a valid non-negative number.';
  if (numeric > 99.99) return 'Tax Rate should be up to 99.99%.';
  if (!hasMaxTwoDecimals(numeric)) return 'Tax Rate should have up to 2 decimal places.';
  return null;
}

export function toNumber(value: string): number {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : 0;
}
